### PR TITLE
Add flag to `CompareWorkspaces` so users can specify `NaN == NaN` behavior

### PR DESCRIPTION
**Pull request by rboston628**
#### Summary of work

Adds a flag to set whether `NaN` values should be considered equal to the `CompareWorkspaces` algorithm.

#### Purpose of work

When `CompareWorkspaces` was originally written, it was written so that if two workspaces had `NaN` for a value, the two `NaN` values would compare as equal.  This is contrary to the IEEE 745 specification, which says that `NaN` compares false against everything, including other `NaN`s, so that `NaN == NaN` should evaluate as false.

In attempting to fix this, it was found there are at least a few tests relying on the old behavior of comparing `NaN`s.  Therefore, this provides a flag to `CompareWorkspaces` so that users can elect for `NaN`s to be considered equal.

See also Issue #38088 

Related to ORNL [EWM 7196](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20%28Change%20Management%29#action=com.ibm.team.workitem.viewWorkItem&id=7196)
  (no further details at that link, only linking for bookkeeping)

#### Further detail of work

Within `CompareWorkspaces`, the methods to check if two values are within a tolerance have been replaced to rely on the `FloatingPointComparison` operators, with a check to return `true` if both values are `NaN` and the `NaNsEqual` flag was set to `true`.

The `CheckAllData` flag will now work on two type of peaks workspaces.

More tests were added for `CompareWorkspaces` and `FloatingPointComparison`, to handle more cases involving `NaN`s or extreme numbers.

A few cppcheck suppressions were removed.

Also, some comments were added in some parts of the code.  These places were found using a search for `abs(` and `fabs(` within the code for possible replacements with `withinAbsoluteDifference`.  While not suitable for replacements, they could lead to improvements in the code.

It was necessary to fix several system tests, which relied on previous behavior for `NaN` comparison.  When the `NaN`s were only in the uncertainties, then checking uncertainty was turned off.  Otherwise, with `NaN`s in the `y`-values. the `NaNsEqual` flag was set for the comparison.

The following tests were updated to work with new `NaN` behavior:
- ISISIndirectInelastic.OSIRISMultiFileSummedReduction
- ISIS_PowderPolarisTest.FocusTestAbsorptionPaalmanPings
- DirectILLAutoProcessTest.DirectILLAuto_PANTHER_Powder_Test
- DirectILLAutoProcessTest.DirectILLAuto_SHARP_Powder_Test
- MDNormHYSPECTest.MDNormHYSPECTest
- SANSBeamCentreFinderCoreTest.SANSBeamCentreCoreRunnerTest
- SANSReductionCoreTest.SANSReductionCoreRunnerTest
- SANSILLReductionTest.ILL_D33_LTOF_Test
- D7AbsoluteCrossSectionsTest.ILL_D7_TimeOfFlight_Test
- ReflectometryQuickMultiDetector.ReflectometryQuickMultiDetector
- ILLDirectGeometryReductionTest.IN4
- ILLDirectGeometryReductionTest.IN5_Mask_Non_Overlapping_Bins
- FlipperEfficiencyTest.FlipperEfficiencyPolarisedTest
- FlipperEfficiencyTest.FlipperEfficiencyUnpolarisedTest
- AbinsTest.AbinsCRYSTAL2D
- SANSILLAutoProcessTest.D11_AutoProcess_IQxQy_Test
- SANSILLAutoProcessTest.D11_AutoProcess_Multiple_Transmissions_Test
- SANSILLAutoProcessTest.D33_AutoProcess_IPhiQ_Test
- ISISIndirectEnergyTransferTest.ISISIndirectEnergyTransferTest
- MuonProcessTest.MuonProcessTest

The following three tests relied on comparing `NaN`s to finite floating points, which used to evaluate as true.  There is no way to fix these without fixing the reference files.  They are marked to skip, and in their previous forms were not meaningful.  It is possible replacing the `NaN`s with `0` could fix some of them.
- CNCSReductionTest.CNCSReductionTest
- SANSZOOMTomlFileConversion.ZOOMV1UserFileTest_m8
- ARCSReductionTest.ARCSReductionTest

See also issue #38088

### To test:

New tests were added to
- `CompareWorkspacesTest.h`
- `TableColumnTest.h`
- `FloatPointComparisonTest.h`
 
covering several situations with `NaN`s and also with extreme values.

These tests should be convincing that the behavior of `NaN`s with and without the flag is as expected.

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: Framework/Algorithms/src/CompareWorkspaces.cpp [hunk 1]
```diff
@@ -634,7 +633,7 @@ bool CompareWorkspaces::checkData(const API::MatrixWorkspace_const_sptr &ws1,
 
   // Now check the data itself
   PARALLEL_FOR_IF(m_parallelComparison && ws1->threadSafe() && ws2->threadSafe())
-  for (long i = 0; i < static_cast<long>(numHists); ++i) {
+  for (std::size_t i = 0; i < numHists; ++i) {
```
[the remark this answers is not in the document]
Comment (rboston628) [reply Framework/Algorithms/src/CompareWorkspaces.cpp:636]: Yeah, turns out I just assumed compilers would know how numbers work.  I'll have to change this back.

File: Framework/Algorithms/src/CompareWorkspaces.cpp [hunk 2]
```diff
@@ -71,23 +72,18 @@ int compareEventLists(Kernel::Logger &logger, const EventList &el1, const EventL
...
-    bool diffweight = false;
-    if (std::abs(e1.pulseTime().totalNanoseconds() - e2.pulseTime().totalNanoseconds()) > tolPulse) {
-      diffpulse = true;
-      ++numdiffpulse;
-    }
-    if (std::abs(e1.tof() - e2.tof()) > tolTof) {
-      difftof = true;
-      ++numdifftof;
-    }
+    bool diffpulse =
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/CompareWorkspaces.cpp:75]: Would it be better to use [`EventList::equals()`](https://github.com/mantidproject/mantid/blob/main/Framework/DataObjects/src/EventList.cpp#L631) for this? The message wouldn't be as clear.

File: Framework/DataObjects/inc/MantidDataObjects/TableColumn.h [hunk 3]
```diff
@@ -64,10 +65,10 @@ template <class Type> class TableColumn : public API::Column {
     std::string name = std::string(typeid(Type).name());
     if ((name.find('i') != std::string::npos) || (name.find('l') != std::string::npos) ||
         (name.find('x') != std::string::npos)) {
-      if (length == 4) {
```
Comment (peterfpeterson) [inline Framework/DataObjects/inc/MantidDataObjects/TableColumn.h:67]: There is a lot of work we could do to make TableColumn not so ugly. Example: move implementation into the cpp file. That is a job for a different PR.

File: Framework/DataObjects/inc/MantidDataObjects/TableColumn.h [hunk 4]
```diff
@@ -282,8 +287,8 @@ inline bool TableColumn<std::string>::compareVectors(const std::vector<std::stri
 
 /// Template specialisation for strings for comparison
 template <>
-inline bool TableColumn<API::Boolean>::compareVectors(const std::vector<API::Boolean> &newVector,
-                                                      double tolerance) const {
+inline bool TableColumn<API::Boolean>::compareVectors(const std::vector<API::Boolean> &newVector, double tolerance,
+                                                      bool const) const {
   (void)tolerance;
```
Comment (peterfpeterson) [inline Framework/DataObjects/inc/MantidDataObjects/TableColumn.h:292]: This isn't in your change already, but we have a macro `UNUSED_ARG` defined in `DllConfig.h` to use. The code would be
```cpp
UNUSED_ARG(tolerance);
```

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Fairly minor change requests. I like the centralized method and making whether nan should be equal to nan is an excellent thing to make explicit.

Later on, we should consider modernizing `TableColumn`.

File: Framework/Algorithms/src/CompareWorkspaces.cpp (hunk 2 above)
Comment (rboston628) [reply Framework/Algorithms/src/CompareWorkspaces.cpp:75]: I don't know enough about either this block of code or `EventList::equals()`, I just saw that this was inefficient and felt compelled to meddle with it.  I'll see if that can serve as a replacement

File: Framework/DataObjects/inc/MantidDataObjects/TableColumn.h (hunk 4 above)
Comment (rboston628) [reply Framework/DataObjects/inc/MantidDataObjects/TableColumn.h:292]: Just figured out what this meant.  I didn't see the `(void)tolerance;` It's actually possible to remove the name `tolerance` from the argument list entirely and not require either `(void)` or `UNUSED_ARG`

File: Framework/Algorithms/src/CompareWorkspaces.cpp (hunk 2 above)
Comment (rboston628) [reply Framework/Algorithms/src/CompareWorkspaces.cpp:75]: Looking at the code, this method is called from another method `compareEventListsInDetail` (L994), which suggests the purpose of this function is precisely the more detailed error messages.  So I think this needs to stay, unfortunately.

**Review by peterfpeterson — APPROVED** [review]

**Pull request merged**